add setting to override available R interpreters

sharon-wang opened the pull request:
### Summary
- addresses https://github.com/posit-dev/positron/issues/6206
- introduces new setting `positron.r.interpreters.override`
    - takes precedence over over the `positron.r.interpreters.exclude`, `positron.r.customBinaries` and `positron.r.customRootFolders` settings

#### Settings UI

<img width="885" alt="image" src="https://github.com/user-attachments/assets/d57335c7-c88a-4b78-a171-12152a7d22a5" />

#### Settings JSON

```json
    "positron.r.interpreters.override": [
        "/opt/local/R/4.3-arm64/Resources/bin/R"
    ]
```

#### Debug Command
`r.interpreters.settingsInfo`

```js
2025-03-14 11:02:50.252 [info] =====================================================================
2025-03-14 11:02:50.252 [info] =============== [START] R INTERPRETER SETTINGS INFO =================
2025-03-14 11:02:50.252 [info] =====================================================================
2025-03-14 11:02:50.252 [info] R interpreter settings: {
  "interpreters.override": [],
  "interpreters.exclude": [],
  "customRootFolders": [
    "/Users/sashimi/scratch"
  ],
  "customBinaries": []
}
2025-03-14 11:02:50.252 [info] =====================================================================
2025-03-14 11:02:50.252 [info] ================ [END] R INTERPRETER SETTINGS INFO ==================
2025-03-14 11:02:50.252 [info] =====================================================================
```

### Release Notes

#### New Features

- The set of available R interpreters can now be overridden via settings (#6206)

#### Bug Fixes

- N/A


### QA Notes

No additional notes!

@:interpreters

sharon-wang committed:
implement `positron.r.interpreters.override` handling

sharon-wang committed:
link to other options in override setting description

sharon-wang commented on extensions/positron-r/package.json:
```diff
@@ -217,6 +217,15 @@
             "markdownDescription": "%r.configuration.interpreters.exclude.markdownDescription%"
           },
+          "positron.r.interpreters.override": {
+            "scope": "window",
```
the scope will be updated to `machine` with all of the other settings in a separate PR

sharon-wang replied:
(Yes that's right!)

It feels like these interpreter settings should all be `machine`-scoped as they refer to specific paths on a given machine. But I think I am more opinionated in that the settings should all have the same scope, and less opinionated about the specific scope they should have.

These are the relevant scopes to choose from, for all these interpreter settings I think! From [VS Code docs](https://code.visualstudio.com/api/references/contribution-points#:~:text=machine%20%2D%20Machine%20specific%20settings%20that%20can%20be%20set%20only%20in%20user%20settings%20or%20only%20in%20remote%20settings.%20For%20example%2C%20an%20installation%20path%20which%20shouldn't%20be%20shared%20across%20machines):
> `machine` - Machine specific settings that can be set only in user settings or only in remote settings. For example, an installation path which shouldn't be shared across machines.
>`machine-overridable` - Machine specific settings that can be overridden by workspace or folder settings.
>`window` - Windows (instance) specific settings which can be configured in user, workspace, or remote settings.

To pin a version of R for a single project sounds like #2659. I feel like the `override` setting is mainly for filtering down the list of available interpreters for admin purposes. Although pinning a version for R could be achieved by specifying a specific interpreter in the `override` option, it feels like a slightly different affordance to me. I am finding it to be intertwined with the Affiliated Runtime concept, which currently only exists on a user's machine and isn't configurable via settings. Perhaps there could be a separate, workspace-level way to configure the affiliated runtime?

Maybe @melissa-barca can comment a bit more on how Workbench Admins may apply user settings for subgroups of users? Is there appetite for any/all of interpreter includes/excludes/overrides/default interpreter path being set at the Workspace level? In other words, should it be easy for a user to supersede these settings at the Workspace level if a Workbench Admin has configured defaults?

jennybc replied:
> Although pinning a version for R could be achieved by specifying a specific interpreter in the override option, it feels like a slightly different affordance to me.

Yeah I agree it almost feels like off-label usage. I'm happy with your existing plan to switch these all to `machine` scope in a later PR and let #2659 continue to track the matter of whether/how to offer such control in a workspace.

melissa-barca replied:
On the Workbench side we're hoping to give admins the ability to configure different profiles for users by taking advantage of the Code OSS profiles: https://github.com/rstudio/rstudio-pro/issues/7468
This would be in place of the user profiles admins can currently configure for RStudio sessions: https://docs.posit.co/ide/server-pro/rstudio_pro_sessions/user_and_group_profiles.html#creating-profiles

I'm going to work on getting an exploratory document typed up about this over the next few days, but I think this is more along the lines of what you're looking for @jennybc ?

jennybc replied:
Sounds good!

jennybc commented on extensions/positron-r/src/r-installation.ts:
```diff
@@ -336,12 +336,51 @@ function getExcludedInstallations(): string[] {
+export function getInterpreterOverridePaths(): string[] {
+	const config = vscode.workspace.getConfiguration('positron.r');
+	const interpretersOverride = config.get<string[]>('interpreters.override') ?? [];
+	if (interpretersOverride.length > 0) {
```
Very weakly held opinion: I think I'd check for length zero and get out early. I'm just a big fan of early exits.

jennybc reviewed (commented):
I'd describe this as consulting the overrides inside of some of our existing sub-strategies for R discovery. Which sort of sprinkles this logic around / pushes it down into helpers.

Did you consider solving from the other direction? Meaning: If overrides are specified, just load up those binaries and bypass all the other sub-strategies. As it stands, it's not obvious to me that we're skipping over, e.g., the registry search and ad hoc binaries. (I realize the runtime discoverer is not really organized favorably for this approach right now, but perhaps we should at least discuss a future refactoring that makes things more obvious.)

sharon-wang committed:
reorganize discovery and move R interpreter settings utils to a separate file

sharon-wang commented:
> I'd describe this as consulting the overrides inside of some of our existing sub-strategies for R discovery. Which sort of sprinkles this logic around / pushes it down into helpers.
> 
> Did you consider solving from the other direction? Meaning: If overrides are specified, just load up those binaries and bypass all the other sub-strategies. As it stands, it's not obvious to me that we're skipping over, e.g., the registry search and ad hoc binaries. (I realize the runtime discoverer is not really organized favorably for this approach right now, but perhaps we should at least discuss a future refactoring that makes things more obvious.)

I was aiming for minimal lift in my original approach, which has the effect you described. After looking at the code, I think the reorganization in https://github.com/posit-dev/positron/pull/6655/commits/37adb983720746406f95ffe14cf2a2a59dc08442 might be ok! I moved a bunch of things around, so happy to do a virtual "desk-side" review if that's easier!

I reorganized things to favour the overrides and skip over everything else if provided. I also moved the interpreter settings code to a separate file.

sharon-wang committed:
handle some sad path cases

- no interpreters found in override locations
- override locations result in duplicate binaries discovered

jennybc commented on extensions/positron-r/src/provider.ts:
```diff
@@ -46,70 +52,15 @@ export enum RRuntimeSource {
-		rInstallations.push(new RInstallation(currentBinary.path, true, currentBinary.reasons));
-		binaries.delete(currentBinary.path);
-	}
+	// Discover R binaries on the system
```
It feels like we gained several unnecessary comments (I think I recognize this behaviour from when I work with co-pilot). And I think we lost some actually useful comments also. I'll point those out.

sharon-wang replied:
I added the comments in deliberately, but I have a bias for verbosity and using comments as sort of a "step heading" for chunks of code 😆 I removed some of them and restored any that were missing post-helper-refactor!

jennybc commented on extensions/positron-r/src/provider.ts:
```diff
@@ -46,70 +52,15 @@ export enum RRuntimeSource {
+	const rInstallations: RInstallation[] = binaries
+		.map(rbin => new RInstallation(rbin.path, rbin.path === currentBinary, rbin.reasons))
 		.filter(r => {
+			// Filter out any rejected R installations
```
Another example of a comment that doesn't seem necessary. I won't point all of these out as it's somewhat subjective, but it would be great if you took a pass at removing some (I think) co-pilot added. In some cases, the code seems pretty self-explanatory already and in others, the new comment feels redundant with more details, in-place comments just below (eg. sorting by version number).

jennybc commented on extensions/positron-r/src/provider.ts:
```diff
@@ -181,6 +135,57 @@ export async function* rRuntimeDiscoverer(): AsyncGenerator<positron.LanguageRun
+		return { binaries: overrideBinaries, currentBinary: undefined };
+	}
+
+	// Consult various sources of R binaries
```
Here's where I feel like some helpful comments have actually been removed, when this got extracted into a helper (I'm supportie of the helper).

jennybc replied:
Update: I now see that some of this got transferred to the docstring above the helper. So now I'm just asking that you conform that any wisdom in those comments got preserved and landed in what seems like the most useful place for code maintenance.

sharon-wang committed:
add command `r.interpreters.settingsInfo` to log R interpreters settings to output

jennybc commented:
I definitely think this refactoring, inspired by the need to insert the overrides, is a good move!

melissa-barca reviewed (dismissed):
LGTM!

sharon-wang committed:
comment pruning and gardening

sharon-wang commented:
Latest changes:

- additional handling for a few more override scenarios [a732f81](https://github.com/posit-dev/positron/pull/6655/commits/a732f81a520bfa1a24cc5422feee1327963376f3)
- I've added in a debugging command `r.interpreters.settingsInfo` which prints the interpreter settings to R Language Pack Output [59a1f19](https://github.com/posit-dev/positron/pull/6655/commits/59a1f19472192a16b60fcb159b1b2028d958f02f) (I added a similar one for Python already)
- some comment pruning and gardening [d7a43ba](https://github.com/posit-dev/positron/pull/6655/commits/d7a43ba8d2558671333042d4dbf3a739cfd338dc)

Will merge once checks are green!
